registry/storage/driver/s3-aws/s3.go

```diff
@@ -665,103 +665,117 @@ func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.Read
-					StorageClass:         d.getStorageClass(),
-				})
-				if err != nil {
-					return nil, err
-				}
-				return d.newWriter(ctx, key, *resp.UploadId, nil), nil
-			}
-			return nil, storagedriver.Error{
-				DriverName: driverName,
-				Detail:     fmt.Errorf("append to zero-size path %s unsupported", path),
-			}
+// createMultiPartUpload creates a new multipart upload for the associated
+// bucket and the given key.
+func (d *driver) createMultipartUpload(ctx context.Context, key string) (*string, error) {
+	res, err := d.S3.CreateMultipartUploadWithContext(ctx, &s3.CreateMultipartUploadInput{
+		Bucket:               aws.String(d.Bucket),
+		Key:                  aws.String(key),
+		ContentType:          d.getContentType(),
+		ACL:                  d.getACL(),
+		ServerSideEncryption: d.getEncryptionMode(),
+		SSEKMSKeyId:          d.getSSEKMSKeyID(),
+		StorageClass:         d.getStorageClass(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res.UploadId, nil
+}
+
+// inProgressUpload finds an in-progress multipart upload for the given path.
+// If there is no in-progress upload, a nil upload ID and no error is returned.
+func (d *driver) inProgressUpload(ctx context.Context, path string) (uploadID *string, err error) {
+	var empty bool
+	if err := d.S3.ListMultipartUploadsPagesWithContext(ctx, &s3.ListMultipartUploadsInput{
+		Bucket: aws.String(d.Bucket),
+		Prefix: aws.String(d.s3Path(path)),
+	}, func(page *s3.ListMultipartUploadsOutput, lastPage bool) bool {
+		// This condition is only valid for the first page. Subsequent
+		// pages are guaranteed not to be empty.
+		if len(page.Uploads) == 0 {
+			empty = true
+			return false
 		}
 
-		var allParts []*s3.Part
-		for _, multi := range resp.Uploads {
-			if key != *multi.Key {
-				continue
+		for _, upload := range page.Uploads {
+			if *upload.Key == d.s3Path(path) {
+				uploadID = upload.UploadId
+				return false
 			}
+		}
+		return true
+	}); err != nil {
+		return nil, fmt.Errorf("list multipart uploads pages: %w", err)
+	}
 
-			partsList, err := d.S3.ListPartsWithContext(ctx, &s3.ListPartsInput{
-				Bucket:   aws.String(d.Bucket),
-				Key:      aws.String(key),
-				UploadId: multi.UploadId,
-			})
-			if err != nil {
-				return nil, parseError(path, err)
-			}
-			allParts = append(allParts, partsList.Parts...)
-			for *partsList.IsTruncated {
-				partsList, err = d.S3.ListPartsWithContext(ctx, &s3.ListPartsInput{
-					Bucket:           aws.String(d.Bucket),
-					Key:              aws.String(key),
-					UploadId:         multi.UploadId,
-					PartNumberMarker: partsList.NextPartNumberMarker,
-				})
-				if err != nil {
-					return nil, parseError(path, err)
-				}
-				allParts = append(allParts, partsList.Parts...)
-			}
-			return d.newWriter(ctx, key, *multi.UploadId, allParts), nil
+	if !empty {
+		if uploadID != nil {
+			return uploadID, nil
 		}
+		return nil, storagedriver.PathNotFoundError{Path: path}
```

Comment:
Just some questions about this behavior here - this is basically saying that if there is no multipart upload at the path, but there were results then a `PathNotFoundError` should be returned? Is it common for there to be multiple uploads at the path which don't exactly match the request key? Why is it fine if there aren't already in-progress uploads? 

Should the logic change to:

```go
if !empty && uploadID != nil
```

Maybe I'm missing something? If I am, then I'd like to document it!

Reply:
> Just some questions about this behavior here - this is basically saying that if there is no multipart upload at the path, but there were results then a PathNotFoundError should be returned?

IIRC this handles the situation where upload purging is cleaning up failed uploads 🤔